README.rst
=======
Off-key
=======


.. image:: https://img.shields.io/pypi/v/offkey.svg
        :target: https://pypi.python.org/pypi/offkey

.. image:: https://img.shields.io/travis/harmony-ek/offkey.svg
        :target: https://travis-ci.org/harmony-ek/offkey

.. image:: https://readthedocs.org/projects/offkey/badge/?version=latest
        :target: https://offkey.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status


.. image:: https://pyup.io/repos/github/harmony-ek/offkey/shield.svg
     :target: https://pyup.io/repos/github/harmony-ek/offkey/
     :alt: Updates



Off-key detects and reports anomalies in Harmony networks.


* Free software: MIT license
* Documentation: https://offkey.readthedocs.io.


Features
--------

* TODO

Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.0.1 (2019-03-04)
------------------

* First release on PyPI.


tests/test_offkey.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `offkey` package."""

import pytest

from click.testing import CliRunner

from offkey import offkey
from offkey import cli


@pytest.fixture
def response():
    """Sample pytest fixture.

    See more at: http://doc.pytest.org/en/latest/fixture.html
    """
    # import requests
    # return requests.get('https://github.com/audreyr/cookiecutter-pypackage')


def test_content(response):
    """Sample pytest test function with the pytest fixture as an argument."""
    # from bs4 import BeautifulSoup
    # assert 'GitHub' in BeautifulSoup(response.content).title.string


def test_command_line_interface():
    """Test the CLI."""
    runner = CliRunner()
    result = runner.invoke(cli.main)
    assert result.exit_code == 0
    assert 'offkey.cli.main' in result.output
    help_result = runner.invoke(cli.main, ['--help'])
    assert help_result.exit_code == 0
    assert '--help  Show this message and exit.' in help_result.output


offkey/cli.py
# -*- coding: utf-8 -*-

"""Console script for offkey."""

from base64 import b64encode
from collections import OrderedDict
from hashlib import sha256
from itertools import chain
import json
import sys
import time
import hashlib

import click
from elasticsearch import Elasticsearch
import requests

from .metrics import build_bucket_search, build_result_buckets
from .util import flatten

HOST_AXES = ('cloud.region', 'cloud.instance.id',)

CONFIGS = [
    dict(
        module='system', metricset='cpu',
        metrics={
            '.total.pct:avg': dict(
                info=90, warning=95, error=97, critical=99,
                message='High CPU percentage (1-minute average)',
            )
        },
        window=60, period=10,
    ),
    dict(
        module='system', metricset='cpu',
        metrics={
            '.total.pct:avg': dict(
                info=80, warning=90, error=95, critical=98,
                message='High CPU percentage (5-minute average)',
            )
        },
        window=300, period=60,
    ),
    dict(
        module='system', metricset='cpu',
        metrics={
            '.total.pct:avg': dict(
                info=70, warning=80, error=90, critical=95,
                message='High CPU percentage (15-minute average)',
            )
        },
        window=900, period=60,
    ),
    dict(
        module='system', metricset='load',
        metrics={
            '.norm.1:max': dict(
                info=10, warning=20, error=30, critical=40,
                message='High 1-minute CPU load',
            ),
            '.norm.5:max': dict(
                info=5, warning=10, error=15, critical=20,
                message='High 5-minute CPU load',
            ),
            '.norm.15:max': dict(
                info=2.5, warning=5, error=7.5, critical=10,
                message='High 15-minute CPU load',
            ),
        },
        window=60, period=30,
    ),
    dict(
        module='system', metricset='memory',
        metrics={
            '.used.pct:avg': dict(
                info=80, warning=90, error=95, critical=98,
                message='High memory use',
            )
        },
        window=60, period=30,
    ),
    dict(
        module='system', metricset='diskio', axes=['.name'],
        metrics={
            '.iostat.busy': dict(
                info=90, warning=95, error=97, critical=99,
                message='High 1-minute disk I/O',
            )
        },
        window=60, period=60,
    ),
    dict(
        module='system', metricset='diskio', axes=['.name'],
        metrics={
            '.iostat.busy': dict(
                info=80, warning=90, error=95, critical=97,
                message='High 5-minute disk I/O',
            )
        },
        window=300, period=60,
    ),
    dict(
        module='system', metricset='diskio', axes=['.name'],
        metrics={
            '.iostat.busy': dict(
                info=80, warning=90, error=95, critical=97,
                message='High 15-minute disk I/O',
            )
        },
        window=900, period=60,
    ),
    dict(
        module='system', metricset='process',
        axes=['.pid', '.cpu.start_time', '.name'],  # for PID recycling
        metrics={
            '.cpu.total.pct': dict(
                info=0.001, warning=0.002, error=0.005, critical=0.01,
                message='High process CPU usage',
            )
        },
        window=60, period=30,
    ),
]


# TODO - honor period
# TODO - add differential/rate support (network needs it)
# TODO - support class-based config (ex: diff. throughput per instance type)


def strictly_increasing(L):
    """Return whether the given sequence is strictly increasing."""
    return all(x < y for x, y in zip(L, L[1:]))


def strictly_decreasing(L):
    """Return whether the given sequence is strictly decreasing."""
    return all(x > y for x, y in zip(L, L[1:]))


def _check_val_against_thresh(val, thresh_config):
    tl = []
    for level in 'info', 'warning', 'error', 'critical':
        thresh = thresh_config.get(level)
        if thresh is not None:
            tl.append((level, thresh))
    if not tl:
        # No thresholds, nothing to do
        return None, None
    tvs = [t[1] for t in tl]
    si = strictly_increasing(tvs)
    sd = strictly_decreasing(tvs)
    assert si or sd, (
        f"{thresh_config!r} is neither strictly increasing nor "
        f"strictly decreasing"
    )
    # caveat emptor: si == sd == True if len(tl) == 1
    severity = None
    ref = None
    for level, thresh in tl:
        if (si and val < thresh) or (not si and val > thresh):
            break
        severity, ref = level, thresh
    if severity is None:
        return None, None
    if si:
        diag = f'{val} >= {ref}'
    else:
        diag = f'{val} <= {ref}'
    return severity, diag


@click.command()
def main(args=None):
    """Console script for offkey."""
    from pprint import pprint
    es = Elasticsearch(port=19201, scheme='https', verify_certs=False)
    # TODO ek – quick-and-dirty impl; use asyncio to parallelize
    for config in CONFIGS:
        module = config['module']
        metricset = config['metricset']
        def full(s):
            if s.startswith('.'):
                return f'{module}.{metricset}{s}'
            return s
        axes = tuple(config.get('axes', ()))
        full_axes = HOST_AXES + axes
        metric_configs = config['metrics']
        metrics = tuple(metric_configs.keys())
        window = config['window']
        # period = search_config['period']  # TODO use when looped
        q = build_bucket_search(metrics, axes=full_axes,
                                metricset=metricset, module=module,
                                window=window)
        r = q.using(es).execute()
        t = build_result_buckets(r, metrics=metrics, axes=full_axes)
        for k, v in flatten(t):
            assert len(k) >= len(HOST_AXES) + 1
            host_axes = k[:len(HOST_AXES)]
            metricset_axes = k[len(HOST_AXES):-1]
            metric_spec = k[-1]
            full_metric_spec = full(metric_spec)
            assert len(axes) == len(metricset_axes)
            metric_config = metric_configs[metric_spec]
            severity, diag = _check_val_against_thresh(v, metric_config)
            host_axes_items = zip(map(full, HOST_AXES), host_axes)
            metricset_axes_items = zip(map(full, axes), metricset_axes)
            all_items = sorted(chain(
                host_axes_items,
                metricset_axes_items,
                ((full_metric_spec, None),),
            ))
            key = b64encode(sha256(json.dumps(all_items).encode()).digest()).decode()
            body = dict(dedup_key=key)
            if severity is None:
                body.update(event_action='resolve')
            else:
                now = time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime())
                custom_details = {}
                custom_details.update(host_axes_items)
                custom_details.update(metricset_axes_items)
                custom_details[full_metric_spec] = v
                source = '/'.join(str(axis) for axis in host_axes)
                component = '/'.join(str(axis) for axis in metricset_axes)
                message = metric_config.get('message', 'Metric out of range')
                payload = {
                    'summary': f'{message}: {full_metric_spec} {diag}',
                    'source': source,
                    'severity': severity,
                    'timestamp': now,
                    'component': component,
                    'class': message,
                    'custom_details': custom_details,
                }
                body.update(event_action='trigger', payload=payload)
            pprint(body)
            rk = 'ce942cb5245643edb20b6dc56a6e14f1'  # XXX fixthis
            res = requests.post('https://events.pagerduty.com/v2/enqueue',
                                json=body, headers={'X-Routing-Key': rk})
            try:
                res.raise_for_status()
            except Exception as e:
                print(res.content)
    return 0


if __name__ == "__main__":
    import logging; logging.basicConfig(level=logging.DEBUG)
    sys.exit(main())  # pragma: no cover


offkey/util.py
"""Utilities for offkey."""

import asyncio
from collections.abc import Sequence, Mapping


class Ticker:
    """
    Yield periodic tick times.

    The first tick is due at *start_time*.  Subsequent ticks are due at
    multiples of the given *period*.

    :param start_time: the start time; defaults to the current time.
    :param period: the period at which to yield ticks.
    """

    def __init__(self, *poargs, start_time=None, period, **kwargs):
        """Initialize this instance."""
        super().__init__(*poargs, **kwargs)
        if start_time is None:
            start_time = asyncio.get_running_loop().time()
        self.__start_time = start_time
        self.__period = period
        self.__ticks = 0
        self.__canceled = False

    @property
    def start_time(self):
        """Return the start time."""
        return self.__start_time

    @property
    def period(self):
        """Return the tick period."""
        return self.__period

    @property
    def ticks(self):
        """Return the number of ticks served so far."""
        return self.__ticks

    @property
    def canceled(self):
        """Return whether the ticker has been canceled."""
        return self.__canceled

    @property
    def last_tick(self):
        """Return the last served tick timestamp.

        Return `None` if none was served yet.
        """
        if self.__ticks == 0:
            return None
        return self.__start_time + (self.__ticks - 1) * self.__period

    @property
    def next_tick(self):
        """Return the next tick timestamp to be served; `None` if canceled."""
        if self.__canceled:
            return None
        return self.__start_time + self.__ticks * self.__period

    def cancel(self):
        """Cancel the ticker; make it generate no more ticks."""
        self.__canceled = True

    def __aiter__(self):
        """Return an async iterator."""
        return self

    async def __anext__(self):
        """Return the next element."""
        if self.__canceled:
            raise StopAsyncIteration
        get_time = asyncio.get_running_loop().time
        now = get_time()
        while now < self.next_tick:
            await asyncio.sleep(self.next_tick - now)
            now = get_time()
        self.__ticks += 1
        return self.last_tick


def flatten(o: object):
    """Flatten the given object."""
    yield from _flatten((), o)


def _flatten(pfx, o):
    if isinstance(o, Mapping):
        for k, v in o.items():
            yield from _flatten(pfx + (k,), v)
    elif isinstance(o, Sequence):
        for i, v in enumerate(o):
            yield from _flatten(pfx + (i,), v)
    else:
        yield pfx, o


offkey/metrics.py
"""Metricbeat tabulated aggregation query facility.

Example of querying filesystem utilization and free bytes/inodes::

    from elasticsearch import Elasticsearch
    from offkey.metrics import build_bucket_search, build_result_buckets
    from pprint import pprint

    es = Elasticsearch(port=19201, scheme='https', verify_certs=False)
    ['.used.pct:max',
     '.free:min',
     '.free_files:min']
    axes = ['cloud.instance.id', '.mount_point']
    s = build_bucket_search(
        ['.used.pct:max', '.free:min', '.free_files:min'],
        module='system', metricset='filesystem',
        axes=['cloud.instance.id', '.mount_point'])
    r = s.using(es).execute()
    pprint(r.to_dict())  # print Elasticsearch query, in Python form.
    t = build_result_buckets(s.using(es).execute(), metrics,
                             module='system', metricset='filesystem',
                             axes=axes)
    pprint(t)  # print tabulated result

Example result from the above::

    {'i-00b05b0ec0795d08d': {'/': {'.free:min': 7054245888.0,
                                   '.free_files:min': 4154413.0,
                                   '.used.pct:max': 0.178}},
     'i-00cef31d24fb49a27': {'/': {'.free:min': 7078440960.0,
                                   '.free_files:min': 4154415.0,
                                   '.used.pct:max': 0.17500000000000002}},
     'i-00ddbddbe9134c9e3': {'/': {'.free:min': 7096557568.0,
                                   '.free_files:min': 4154416.0,
                                   '.used.pct:max': 0.17300000000000001}},
     'i-02d11a8c7b656af29': {'/': {'.free:min': 7096700928.0,
                                   '.free_files:min': 4154416.0,
                                   '.used.pct:max': 0.17300000000000001}},
     'i-040d9a77c92eddfd4': {'/': {'.free:min': 6956879872.0,
                                   '.free_files:min': 4154409.0,
                                   '.used.pct:max': 0.189}},
     'i-041ff6caee1975e2d': {'/': {'.free:min': 7054585856.0,
                                   '.free_files:min': 4154413.0,
                                   '.used.pct:max': 0.178}},
     'i-04671d193d2beae0d': {'/': {'.free:min': 7096389632.0,
                                   '.free_files:min': 4154416.0,
                                   '.used.pct:max': 0.17300000000000001}},
     'i-048e18bd42bfb3374': {'/': {'.free:min': 7061102592.0,
                                   '.free_files:min': 4154414.0,
                                   '.used.pct:max': 0.177}},
     'i-04d6a23d056891238': {'/': {'.free:min': 7038185472.0,
                                   '.free_files:min': 4154412.0,
                                   '.used.pct:max': 0.179}},
     'i-068038b1a51ff2c0a': {'/': {'.free:min': 7085850624.0,
                                   '.free_files:min': 4154416.0,
                                   '.used.pct:max': 0.17400000000000002}}}
"""

import math
import time
from typing import Sequence, Mapping, Union, Optional

from elasticsearch_dsl import Search

_mem_used_avg = Search(index='metricbeat-*') \
    .filter('term', **{'metricset.module': 'system'}) \
    .filter('term', **{'metricset.name': 'memory'}) \
    .filter('range', **{'@timestamp': {'gte': '2019-03-02T02:00:00Z',
                                       'lt':  '2019-03-02T02:05:00Z'}}) \
    .aggs \
    .bucket('host', 'terms', field='host.name') \
    .bucket('avg', 'avg', field='system.memory.used.pct')


def unix2es(ts: Union[float, int]) -> str:
    """Convert a UNIX timestamp to Elasticsearch-style ISO 8601 timestamp.

    The UTC time zone suffix (``Z``) is appended to the timestamp.

    Fractional second is rounded down to millisecond precision:

    >>> unix2es(3.14159)
    '1970-01-01T00:00:03.141Z'

    Whole seconds (`float` or `int`) get zeroed-out fractional part:

    >>> unix2es(3.0)
    '1970-01-01T00:00:03.000Z'
    >>> unix2es(3)
    '1970-01-01T00:00:03.000Z'

    :param ts: UNIX-style timestamp
    :return: a string in ISO 8601 UTC format with millisecond precision.
    """
    ymdhms = time.strftime('%Y-%m-%dT%H:%M:%S', time.gmtime(ts))
    ms = int((ts - math.floor(ts)) * 1000)
    return f'{ymdhms}.{ms:03d}Z'


DEFAULT_AXES = ('host.name',)
"""Default axes (bucket keys) to use if not specified."""

DEFAULT_AGG = 'avg'
"""Default aggregation to use for metrics specified without one."""

DEFAULT_WINDOW = 300.0
"""Default time window over which to aggregate."""


def build_bucket_search(metrics: Sequence[str],
                        metricset: Optional[str] = None,
                        module: Optional[str] = None,
                        axes: Sequence[str] = DEFAULT_AXES,
                        with_docs: bool = False,
                        window: Optional[float] = None,
                        end_time: Optional[float] = None) -> Search:
    """Build Elasticsearch bucket search for one or more Metricbeat metrics.

    Each metric specification in *metrics* takes the form of ``name[:agg]``.

    ``name`` is the metric name, such as ``system.memory.used.pct``.  ``agg``
    is the bucket aggregation function to use for the metric.

    *metricset* and *module* are the Metricbeat module and metricset names.
    The given *metrics* must all belong to this metricset; those that do
    not are not returned.

    If *metricset* and *module* are not specified, they are taken from the
    first two dotted components of the first metric name, a convention used by
    Metricbeat modules.  For example, ``system.memory.used.pct`` is assumed to
    be in the metricset ``memory`` provided by the ``system`` module.  This is
    useful when querying a single metric in the metricset.

    Conversely, if *metricset* and *module* are specified, metric names can be
    shortened by not having them and starting with a dot instead.  For
    example, if *module*=``system`` and *metricset*=``memory``, the metric
    name ``system.memory.used.pct`` can be shortened as ``.used.pct``.  This
    is useful when querying more than one metrics in the same metricset.

    *axes* are a sequence of aggregation bucket keys.  The default list
    includes only the ``host.name`` standard Metricbeat hostname.  It can be
    extended with more axes, ex: ``('host.name', 'system.network.name')`` for
    network interface table (the `network` metricset).  The host name axis
    can also be replaced by another axis, such as `cloud.instance.id`.
    Axis names can be shortened if the metricset and module names are known
    (given or deduced from *metrics*).

    *window* and *end_time* define the time range, and default to 5
    minutes (300 seconds) and the current UNIX time (on this system, not on
    Elasticsearch), respectively.

    By default, the build query requests Elasticsearch not to return matched
    documents.  *with_docs*=`True` changes this, and requests matched docs.

    :param metrics: Sequence of one or more metric specifications.
    :param metricset: Metricset name, such as `memory`.
    :param module: Module name, such as `system`.
    :param axes: Sequence of axis along which to split the result.
    :param with_docs: Whether to include raw documents matched.
    :param window: Duration over which to aggregate.
    :param end_time: The end time of the aggregation window.
    :return: Elasticsearch DSL query built.
    """
    metric_path = metrics[0].split(':')[0].split('.')
    if module is None:
        module = metric_path[0]
    if metricset is None:
        metricset = metric_path[1]
    if window is None:
        window = DEFAULT_WINDOW
    if end_time is None:
        end_time = time.time()

    def full(name):
        if name.startswith('.'):
            return module + '.' + metricset + name
        return name

    start_time = end_time - window
    s = Search(index='metricbeat-*') \
        .filter('term', **{'metricset.module': module}) \
        .filter('term', **{'metricset.name': metricset}) \
        .filter('range', **{'@timestamp': {'gte': unix2es(start_time),
                                           'lt':  unix2es(end_time)}})
    a = s.aggs
    for axis in axes:
        a = a.bucket(axis, 'terms', field=full(axis))
    for metric_agg in metrics:
        try:
            metric, agg = metric_agg.rsplit(':', 1)
        except ValueError:
            metric, agg = metric_agg, DEFAULT_AGG
        a.bucket(metric_agg, agg, field=full(metric))
    if not with_docs:
        s = s[:0]
    return s


def build_result_buckets(result, metrics: Sequence[str],
                         axes: Sequence[str] = DEFAULT_AXES) -> Mapping:
    """Tabulate execution result of a query built by `build_bucket_search()`.

    :param result: The return value of a `Search.execute()` call.
    :param metrics: See `build_bucket_search()`.
    :param axes: See `build_bucket_search()`.
    :return:
        The tabulated result.  It is a nested dict, with same depth as the
        number of axes, with the leaf dict keyed by metric specs.
    """
    def _brb(aggs: Mapping, remaining_axes: Sequence[str]):
        if not remaining_axes:
            return {metric: aggs[metric].value for metric in metrics}
        axis, remaining_axes = remaining_axes[0], remaining_axes[1:]
        return {
            bucket.key: _brb(bucket, remaining_axes)
            for bucket in aggs[axis]
        }

    return _brb(result.aggs, axes)
